### DISCUSSION TITLE
Get USB Joystick/Gamepad working

### DISCUSSION THREAD

**Discussion by paulscottrobson** [paulscottrobson/neo6502-firmware#192] [Implemented Ideas]

**Comment by astralaster** [comment]
Do you already have an idea on how the API should look like? I think it would be great to not limit the gamepads on a specific button count or similiar.

**Comment by paulscottrobson** [comment]
Early experiments umpromising. Have two joysticks (a faux XBox controller, and a classic joystick), both of which work, neither are detected (e.g. mount does not appear to be called). More work required on this.

**Comment by paulscottrobson** [comment]
> Do you already have an idea on how the API should look like? I think it would be great to not limit the gamepads on a specific button count or similiar.

Well, there can be up to 8 fire buttons supported as is ; the joypad() return value is a bit pattern, so it can be masked or simply used as "any button" by testing for non-zero.

**Comment by astralaster** [comment]
I think to honor the neo in neo6502, it should support a standard xbox like controller. So analog input and all buttons. Current plan is to leave / implement the current existing API for a single controller and also add a more complex one for multiple controllers with more controls.

**Comment by paulscottrobson** [comment]
This is the ideal, preferably it will support anything from a basic single button joystick up to those sorts of controllers ; the backup is probably to use a Nunchunk (which oddly seem to be freely available and cheap) and Olimex's interface board. I haven't tried it yet (I do have one) but I can't see why it wouldn't work.

Probably most of any hardware is going to be via I2C or USB. There was someone wanted to do hardware via the 40 pin bus. I'm not sure this is possible except for output.

**Comment by astralaster** [comment]
Some news: HID class gamepads seem to work just fine. But gamepads like xbox are not HID class, but vendor specific. There might be ways to extend compatibility, but we will first start with just adding HID class gamepad support.

**Comment by astralaster** [comment]
#344 added usb gamepad support.
